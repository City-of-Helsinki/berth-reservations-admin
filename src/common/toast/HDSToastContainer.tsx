import React from 'react';
import { ToastContainer } from 'react-toastify';

import styles from './hdsToastContainer.module.scss';

export const HDSToastContainerId = 'HDSToastContainer';

const HDSToastContainer: React.FC = () => {
  return (
    <ToastContainer
      closeButton={false}
      autoClose={false}
      hideProgressBar={true}
      closeOnClick={false}
      draggable={false}
      className={styles.container}
      containerId={HDSToastContainerId}
      enableMultiContainer={true}
    />
  );
};

export default HDSToastContainer;
