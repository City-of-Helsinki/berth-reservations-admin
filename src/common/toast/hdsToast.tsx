import React, { useEffect } from 'react';
import { Notification, NotificationProps } from 'hds-react';
import { toast } from 'react-toastify';
import { v4 as uuidv4 } from 'uuid';
import { useTranslation } from 'react-i18next';
import { GraphQLError } from 'graphql';

import i18n from '../../locales/i18n';
import { HDSToastContainerId } from './HDSToastContainer';

interface HDSToastArgs {
  autoDismiss?: boolean;
  autoDismissTime?: number;
  type: NotificationProps['type'];
  labelText: string;
  text: string;
  toastId?: string;
  translated?: boolean;
}

interface NotificationWrapperProps extends HDSToastArgs {
  toastId: string;
  translated: boolean;
}

const AUTO_DISMISS_TIME = 3000;

const NotificationWrapper = ({
  autoDismiss,
  autoDismissTime = AUTO_DISMISS_TIME,
  type,
  labelText,
  text,
  toastId,
  translated,
}: NotificationWrapperProps) => {
  const { t } = useTranslation();

  useEffect(() => {
    if (autoDismiss) {
      const timer = setTimeout(() => {
        toast.dismiss(toastId);
      }, autoDismissTime);
      return () => clearTimeout(timer);
    }
  }, [autoDismiss, autoDismissTime, toastId]);

  return (
    <Notification
      type={type}
      dismissible
      closeButtonLabelText={t('toast.closeToast') ?? ''}
      label={translated ? t(labelText) : labelText}
      onClose={() => toast.dismiss(toastId)}
    >
      {translated ? t(text) : text}
    </Notification>
  );
};

const hdsToast = ({
  autoDismiss = true,
  autoDismissTime = AUTO_DISMISS_TIME,
  type,
  labelText,
  text,
  toastId,
  translated = false,
}: HDSToastArgs) => {
  const id = toastId ?? uuidv4();
  return toast(
    <NotificationWrapper
      autoDismiss={autoDismiss}
      type={type}
      labelText={labelText}
      text={text}
      toastId={id}
      translated={translated}
      autoDismissTime={autoDismissTime}
    />,
    {
      toastId: id,
      containerId: HDSToastContainerId,
    }
  );
};

hdsToast.graphQLErrors = (errors: ReadonlyArray<GraphQLError>) => {
  errors.forEach((error) =>
    hdsToast({
      autoDismiss: false,
      type: 'error',
      labelText:
        error.extensions?.code && i18n.exists(`toast.graphQLErrors.${error.extensions.code}.label`)
          ? `toast.graphQLErrors.${error.extensions?.code}.label`
          : 'toast.graphQLErrors.default.label',
      text:
        error.extensions?.code && i18n.exists(`toast.graphQLErrors.${error.extensions.code}.description`)
          ? `toast.graphQLErrors.${error.extensions?.code}.description`
          : 'toast.graphQLErrors.default.description',
      translated: true,
    })
  );
};

export default hdsToast;
