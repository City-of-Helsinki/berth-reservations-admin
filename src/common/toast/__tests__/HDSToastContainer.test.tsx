import React, { useState } from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { wait } from '@apollo/react-testing';
import { act } from 'react-dom/test-utils';

import hdsToast from '../hdsToast';
import HDSToastContainer from '../HDSToastContainer';

describe('HDSToastContainer', () => {
  const TestComponent: React.FC = () => {
    const [count, setCount] = useState(0);
    return (
      <div>
        <HDSToastContainer />
        <button
          id="toastBtn"
          onClick={() => {
            hdsToast({ type: 'error', labelText: 'labelText', text: 'text', toastId: `toast-${count}` });
            setCount(count + 1);
          }}
        />
      </div>
    );
  };

  const getWrapper = () => {
    return mount(<TestComponent />);
  };

  const simulateToast = async (wrapper: ReactWrapper) => {
    wrapper.find('#toastBtn').simulate('click');
    await act(async () => await wait(100));
    wrapper.update();
  };

  const simulateCloseToast = async (wrapper: ReactWrapper, toastId: number) => {
    wrapper.find(`#toast-${toastId}`).find('button[title="common.closeToast"]').simulate('click');
    await act(async () => await wait(1500));
    wrapper.update();
  };

  it('adds toasts to the container when hdsToast is called', async () => {
    const wrapper = getWrapper();
    await simulateToast(wrapper);
    expect(wrapper.find('#toast-0')).toHaveLength(1);
    expect(wrapper.find('.Toastify__toast')).toHaveLength(1);
  });

  it('supports multiple simultaneous toasts', async () => {
    const wrapper = getWrapper();
    await simulateToast(wrapper);
    await simulateToast(wrapper);
    expect(wrapper.find('#toast-0')).toHaveLength(1);
    expect(wrapper.find('#toast-1')).toHaveLength(1);
    expect(wrapper.find('.Toastify__toast')).toHaveLength(2);
  });

  it('closes toast when the notifications close button is clicked', async () => {
    const wrapper = getWrapper();
    await simulateToast(wrapper);
    expect(wrapper.find('#toast-0')).toHaveLength(1);
    await simulateCloseToast(wrapper, 0);
    expect(wrapper.find('#toast-0')).toHaveLength(0);
    expect(wrapper.find('.Toastify__toast')).toHaveLength(0);
  });
});
