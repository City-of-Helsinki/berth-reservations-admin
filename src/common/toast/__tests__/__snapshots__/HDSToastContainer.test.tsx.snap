// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HDSToastContainer adds toasts to the container when hdsToast is called 1`] = `"<div><div class=\\"Toastify\\" id=\\"HDSToastContainer\\"><div class=\\"Toastify__toast-container Toastify__toast-container--top-right container\\"><div id=\\"0\\" class=\\"Toastify__toast Toastify__toast--default Toastify__bounce-enter--top-right\\" style=\\"animation-fill-mode: forwards; animation-duration: 750ms;\\"><div role=\\"alert\\" class=\\"Toastify__toast-body\\"><div style=\\"height: 0px; overflow-y: hidden; position: relative;\\"><div class=\\"Notification-module_notification__2hN3H notification_hds-notification__2DQPN Notification-module_error__2sn9y notification_hds-notification--error__3T1Of\\"><div class=\\"Notification-module_label__1bhNu notification_hds-notification__label__2t1lR\\"><span class=\\"Notification-module_icon__1ans7 notification_hds-icon__29HQM\\" aria-hidden=\\"true\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 30 25\\" class=\\"Notification-module_iconWarning__2bF7u notification_hds-icon--warning__BEw15\\"><g fill-rule=\\"evenodd\\"><path fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2.5\\" d=\\"M16.457 2.108L28.033 20.35a1.8 1.8 0 0 1-1.52 2.765H3.362a1.8 1.8 0 0 1-1.52-2.765l11.576-18.24a1.8 1.8 0 0 1 3.04 0z\\"></path><path d=\\"M16.5 16.916v2.38H14v-2.38h2.5zM16.5 7v7.783H14V7.001L16.5 7z\\"></path></g></svg></span><span class=\\"Notification-module_labelText__1qBt6 helpers_text-bold__1mCMR helpers_text-md__2Td0C\\">labelText</span></div><div class=\\"Notification-module_bodyText__2seBK notification_hds-notification__body__3cobH helpers_text-body__1PMI3\\"><button class=\\"DismissableNotification-module_buttonClose__2HSH1 notification_hds-notification__close-button__1yGIj DismissableNotification-module_error__1lYkg notification_hds-notification--error__3T1Of\\" type=\\"button\\" title=\\"common.closeToast\\" aria-label=\\"common.closeToast\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 15 15\\" class=\\"DismissableNotification-module_iconClose__1JHO3 notification_hds-icon__29HQM\\"><path fill-rule=\\"nonzero\\" d=\\"M13.412 0L15 1.588 9.088 7.5 15 13.412 13.412 15 7.5 9.088 1.588 15 0 13.412 5.911 7.5 0 1.588 1.588 0 7.5 5.912 13.412 0z\\"></path></svg></button>text</div></div></div></div></div></div></div><button id=\\"toastBtn\\"></button></div>"`;

exports[`HDSToastContainer closes toast when the notifications close button is clicked 1`] = `"<div><div class=\\"Toastify\\" id=\\"HDSToastContainer\\"><div class=\\"Toastify__toast-container Toastify__toast-container--top-right container\\"><div id=\\"0\\" class=\\"Toastify__toast Toastify__toast--default Toastify__bounce-enter--top-right\\" style=\\"animation-fill-mode: forwards; animation-duration: 750ms;\\"><div role=\\"alert\\" class=\\"Toastify__toast-body\\"><div style=\\"height: 0px; overflow-y: hidden; position: relative;\\"><div class=\\"Notification-module_notification__2hN3H notification_hds-notification__2DQPN Notification-module_error__2sn9y notification_hds-notification--error__3T1Of\\"><div class=\\"Notification-module_label__1bhNu notification_hds-notification__label__2t1lR\\"><span class=\\"Notification-module_icon__1ans7 notification_hds-icon__29HQM\\" aria-hidden=\\"true\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 30 25\\" class=\\"Notification-module_iconWarning__2bF7u notification_hds-icon--warning__BEw15\\"><g fill-rule=\\"evenodd\\"><path fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2.5\\" d=\\"M16.457 2.108L28.033 20.35a1.8 1.8 0 0 1-1.52 2.765H3.362a1.8 1.8 0 0 1-1.52-2.765l11.576-18.24a1.8 1.8 0 0 1 3.04 0z\\"></path><path d=\\"M16.5 16.916v2.38H14v-2.38h2.5zM16.5 7v7.783H14V7.001L16.5 7z\\"></path></g></svg></span><span class=\\"Notification-module_labelText__1qBt6 helpers_text-bold__1mCMR helpers_text-md__2Td0C\\">labelText</span></div><div class=\\"Notification-module_bodyText__2seBK notification_hds-notification__body__3cobH helpers_text-body__1PMI3\\"><button class=\\"DismissableNotification-module_buttonClose__2HSH1 notification_hds-notification__close-button__1yGIj DismissableNotification-module_error__1lYkg notification_hds-notification--error__3T1Of\\" type=\\"button\\" title=\\"common.closeToast\\" aria-label=\\"common.closeToast\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 15 15\\" class=\\"DismissableNotification-module_iconClose__1JHO3 notification_hds-icon__29HQM\\"><path fill-rule=\\"nonzero\\" d=\\"M13.412 0L15 1.588 9.088 7.5 15 13.412 13.412 15 7.5 9.088 1.588 15 0 13.412 5.911 7.5 0 1.588 1.588 0 7.5 5.912 13.412 0z\\"></path></svg></button>text</div></div></div></div></div></div></div><button id=\\"toastBtn\\"></button></div>"`;

exports[`HDSToastContainer supports multiple simultaneous toasts 1`] = `"<div><div class=\\"Toastify\\" id=\\"HDSToastContainer\\"><div class=\\"Toastify__toast-container Toastify__toast-container--top-right container\\"><div id=\\"0\\" class=\\"Toastify__toast Toastify__toast--default Toastify__bounce-enter--top-right\\" style=\\"animation-fill-mode: forwards; animation-duration: 750ms;\\"><div role=\\"alert\\" class=\\"Toastify__toast-body\\"><div style=\\"height: auto; overflow-y: hidden; position: relative;\\"><div class=\\"Notification-module_notification__2hN3H notification_hds-notification__2DQPN Notification-module_error__2sn9y notification_hds-notification--error__3T1Of\\"><div class=\\"Notification-module_label__1bhNu notification_hds-notification__label__2t1lR\\"><span class=\\"Notification-module_icon__1ans7 notification_hds-icon__29HQM\\" aria-hidden=\\"true\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 30 25\\" class=\\"Notification-module_iconWarning__2bF7u notification_hds-icon--warning__BEw15\\"><g fill-rule=\\"evenodd\\"><path fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2.5\\" d=\\"M16.457 2.108L28.033 20.35a1.8 1.8 0 0 1-1.52 2.765H3.362a1.8 1.8 0 0 1-1.52-2.765l11.576-18.24a1.8 1.8 0 0 1 3.04 0z\\"></path><path d=\\"M16.5 16.916v2.38H14v-2.38h2.5zM16.5 7v7.783H14V7.001L16.5 7z\\"></path></g></svg></span><span class=\\"Notification-module_labelText__1qBt6 helpers_text-bold__1mCMR helpers_text-md__2Td0C\\">labelText</span></div><div class=\\"Notification-module_bodyText__2seBK notification_hds-notification__body__3cobH helpers_text-body__1PMI3\\"><button class=\\"DismissableNotification-module_buttonClose__2HSH1 notification_hds-notification__close-button__1yGIj DismissableNotification-module_error__1lYkg notification_hds-notification--error__3T1Of\\" type=\\"button\\" title=\\"common.closeToast\\" aria-label=\\"common.closeToast\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 15 15\\" class=\\"DismissableNotification-module_iconClose__1JHO3 notification_hds-icon__29HQM\\"><path fill-rule=\\"nonzero\\" d=\\"M13.412 0L15 1.588 9.088 7.5 15 13.412 13.412 15 7.5 9.088 1.588 15 0 13.412 5.911 7.5 0 1.588 1.588 0 7.5 5.912 13.412 0z\\"></path></svg></button>text</div></div></div></div></div><div id=\\"1\\" class=\\"Toastify__toast Toastify__toast--default Toastify__bounce-enter--top-right\\" style=\\"animation-fill-mode: forwards; animation-duration: 750ms;\\"><div role=\\"alert\\" class=\\"Toastify__toast-body\\"><div style=\\"height: auto; overflow-y: hidden; position: relative;\\"><div class=\\"Notification-module_notification__2hN3H notification_hds-notification__2DQPN Notification-module_error__2sn9y notification_hds-notification--error__3T1Of\\"><div class=\\"Notification-module_label__1bhNu notification_hds-notification__label__2t1lR\\"><span class=\\"Notification-module_icon__1ans7 notification_hds-icon__29HQM\\" aria-hidden=\\"true\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 30 25\\" class=\\"Notification-module_iconWarning__2bF7u notification_hds-icon--warning__BEw15\\"><g fill-rule=\\"evenodd\\"><path fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2.5\\" d=\\"M16.457 2.108L28.033 20.35a1.8 1.8 0 0 1-1.52 2.765H3.362a1.8 1.8 0 0 1-1.52-2.765l11.576-18.24a1.8 1.8 0 0 1 3.04 0z\\"></path><path d=\\"M16.5 16.916v2.38H14v-2.38h2.5zM16.5 7v7.783H14V7.001L16.5 7z\\"></path></g></svg></span><span class=\\"Notification-module_labelText__1qBt6 helpers_text-bold__1mCMR helpers_text-md__2Td0C\\">labelText</span></div><div class=\\"Notification-module_bodyText__2seBK notification_hds-notification__body__3cobH helpers_text-body__1PMI3\\"><button class=\\"DismissableNotification-module_buttonClose__2HSH1 notification_hds-notification__close-button__1yGIj DismissableNotification-module_error__1lYkg notification_hds-notification--error__3T1Of\\" type=\\"button\\" title=\\"common.closeToast\\" aria-label=\\"common.closeToast\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 15 15\\" class=\\"DismissableNotification-module_iconClose__1JHO3 notification_hds-icon__29HQM\\"><path fill-rule=\\"nonzero\\" d=\\"M13.412 0L15 1.588 9.088 7.5 15 13.412 13.412 15 7.5 9.088 1.588 15 0 13.412 5.911 7.5 0 1.588 1.588 0 7.5 5.912 13.412 0z\\"></path></svg></button>text</div></div></div></div></div></div></div><button id=\\"toastBtn\\"></button></div>"`;
